### PR TITLE
`azurerm_app_configuration` - omit no permission error in get soft-deleted 

### LINKED ISSUE #19605: Support for disabling detection of soft deleted azurerm_app_configuration resources
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In the current version of the AzureRM provider (3.34.0), running `terraform apply` with a plan that includes new azurerm_app_configuration resources will check for soft deleted App Configuration stores, which requires the subscription-level permission `Microsoft.AppConfiguration/locations/deletedConfigurationStores/read`. Without that permission, apply errors out from the API response's 403 status code.
```
azurerm_app_configuration.default: Creating...
│ Error: checking for presence of deleted Deleted Configuration Store (Subscription: "redacted"
│ Location: "westus"
│ Config Store Name: "redacted"): deletedconfigurationstores.DeletedConfigurationStoresClient#ConfigurationStoresGetDeleted: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'redacted' with object id 'redacted' does not have authorization to perform action 'Microsoft.AppConfiguration/locations/deletedConfigurationStores/read' over scope '/subscriptions/redacted/providers/Microsoft.AppConfiguration/locations/westus/deletedConfigurationStores/redacted' or the scope is invalid. If access was recently granted, please refresh your credentials."
```
We are unable to create custom Azure roles in our tenant (so we can't add the App Configuration purge permissions to our service principals), and we do not want to give our environment-specific service principals the Owner or Contributor roles on the whole subscription. It would be helpful if there was a way to disable the check for soft-deleted App Configuration resources when applying a plan that includes a new one.

### New or Affected Resource(s)/Data Source(s)

azurerm_app_configuration

### Potential Terraform Configuration

```hcl
# Ideally, this already-possible configuration would skip the check for soft deleted App Configuration resources.

terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "3.34.0"
    }
  }
}

provider "azurerm" {
  features {
    app_configuration {
      recover_soft_deleted = false
    }
  }
}

resource "azurerm_app_configuration" "default" {
  name                     = "my-app-config"
  location                 = "westus"
  resource_group_name      = "my-rg"
  sku                      = "standard"
  purge_protection_enabled = false
}
```

### References

https://github.com/hashicorp/terraform-provider-azurerm/issues/18827

### PR DESCRIPTION
**Pull request by teowa**
resolves #19605

Omit no permission error in get soft-deleted.

### CODE REVIEW

**Review by stephybun — CHANGES_REQUESTED** [review]
The feature flag here https://github.com/hashicorp/terraform-provider-azurerm/blob/eb2e4c67f157bd5f752e46a90829228322c5a477/internal/services/appconfiguration/app_configuration_resource.go#L260
Needs to wrap the whole retrieval of the deleted configuration stores which begins on line 249
https://github.com/hashicorp/terraform-provider-azurerm/blob/eb2e4c67f157bd5f752e46a90829228322c5a477/internal/services/appconfiguration/app_configuration_resource.go#L249

File: internal/services/appconfiguration/app_configuration_resource.go [hunk 1]
```diff
@@ -537,7 +538,8 @@ func resourceAppConfigurationDelete(d *pluginsdk.ResourceData, meta interface{})
 		if purgeProtectionEnabled {
 			deletedInfo, err := deletedConfigurationStoresClient.ConfigurationStoresGetDeleted(ctx, deletedId)
 			if err != nil {
-				return fmt.Errorf("retrieving the Deletion Details for %s: %+v", *id, err)
+				log.Printf("[DEBUG] While purging the soft-deleted, retrieving the Deletion Details for %s: %+v", *id, err)
+				return nil
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration_resource.go:542]: This should still return an error.

**Review by stephybun — CHANGES_REQUESTED** [review]

**Comment by teowa** [comment]
@stephybun , thanks for reviewing this. I have modified the code. But I find if move the retrieval of deleted into the features toggle scope, we cannot return the [`optedOutOfRecoveringSoftDeletedAppConfigurationErrorFmt`](https://github.com/hashicorp/terraform-provider-azurerm/pull/19661/files#diff-8edf864f3cdcffb8141e63a4f06b67d86de38abf6473be0779d5ed7ae556c050L653) error when the deleted exists and recover deleted features toggle disabled.

And there is test failure, the App Conf resource creation is failed and should net exists but returned by resource group list API, which might related to #18796

```
TF_ACC=1 go test -v ./internal/services/appconfiguration -parallel 20 -test.run=TestAccAppConfiguration_softDeleteRecoveryDisabled -timeout 1440m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAppConfiguration_softDeleteRecoveryDisabled
=== PAUSE TestAccAppConfiguration_softDeleteRecoveryDisabled
=== CONT  TestAccAppConfiguration_softDeleteRecoveryDisabled
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: deleting Resource Group "acctestRG-appconfig-221214143801923073": the Resource Group still contains Resources.
        
        Terraform is configured to check for Resources within the Resource Group when deleting the Resource Group - and
        raise an error if nested Resources still exist to avoid unintentionally deleting these Resources.
        
        Terraform has detected that the following Resources still exist within the Resource Group:
        
        * `/subscriptions/85b3dbca-5974-4067-9669-67a141095a76/resourceGroups/acctestRG-appconfig-221214143801923073/providers/Microsoft.AppConfiguration/configurationStores/testaccappconf221214143801923073`
        
        This feature is intended to avoid the unintentional destruction of nested Resources provisioned through some
        other means (for example, an ARM Template Deployment) - as such you must either remove these Resources, or
        disable this behaviour using the feature flag `prevent_deletion_if_contains_resources` within the `features`
        block when configuring the Provider, for example:
        
        provider "azurerm" {
          features {
            resource_group {
              prevent_deletion_if_contains_resources = false
            }
          }
        }
        
        When that feature flag is set, Terraform will skip checking for any Resources within the Resource Group and
        delete this using the Azure API directly (which will clear up any nested resources).
        
        More information on the `features` block can be found in the documentation:
        https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#features
        
        
--- FAIL: TestAccAppConfiguration_softDeleteRecoveryDisabled (830.13s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-azurerm/internal/services/appconfiguration      830.144s
FAIL
make: *** [GNUmakefile:99: acctests] Error 1

```

File: internal/services/appconfiguration/app_configuration_resource.go [hunk 2]
```diff
@@ -246,20 +245,20 @@ func resourceAppConfigurationCreate(d *pluginsdk.ResourceData, meta interface{})
...
-	if !response.WasNotFound(deleted.HttpResponse) && !response.WasStatusCode(deleted.HttpResponse, http.StatusForbidden) {
-		if !meta.(*clients.Client).Features.AppConfiguration.RecoverSoftDeleted {
-			return fmt.Errorf(optedOutOfRecoveringSoftDeletedAppConfigurationErrorFmt(name, location))
+	if meta.(*clients.Client).Features.AppConfiguration.RecoverSoftDeleted {
+		deletedConfigurationStoresId := deletedconfigurationstores.NewDeletedConfigurationStoreID(subscriptionId, location, name)
+		deleted, err := deletedConfigurationStoresClient.ConfigurationStoresGetDeleted(ctx, deletedConfigurationStoresId)
+		if err != nil {
+			if !response.WasNotFound(deleted.HttpResponse) {
+				return fmt.Errorf("checking for presence of deleted %s: %+v", deletedConfigurationStoresId, err)
+			}
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration_resource.go:256]:
```suggestion
		if err != nil {
			if response.WasStatusCode(deleted.HttpResponse, http.StatusForbidden) {
				return fmt.Errorf(userIsMissingRelevantPermission(name, location))
			}
			if !response.WasNotFound(deleted.HttpResponse) {
				return fmt.Errorf("checking for presence of deleted %s: %+v", deletedConfigurationStoresId, err)
			}
```

**Review by stephybun — CHANGES_REQUESTED** [review]
We should still return an error if the user is missing the necessary permissions for retrieval and is opted into this feature which is enabled by default, the error message should be changed to something like this:

```
func userIsMissingNecessaryPermission(name, location string) string {
	return fmt.Sprintf(`
An existing soft-deleted App Configuration exists with the Name %q in the location %q, however
the credentials Terraform is using has insufficient permissions to check for an existing soft-deleted App Configuration.
You can opt out of this behaviour by using the "features" block (located within the "provider" block) - more information
can be found here:
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#features
Alternatively you can manually recover this (e.g. using the Azure CLI) and then import
this into Terraform via "terraform import".
`, name, location)
}
```

File: internal/services/appconfiguration/app_configuration_resource.go (hunk 2 above)
Comment (teowa) [reply internal/services/appconfiguration/app_configuration_resource.go]: Thanks for reviewing, I have added the `userIsMissingRelevantPermission` error.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @teowa, there's a test failure for this service, would you mind fixing it? Once that's done this should be good to go!
```
------- Stdout: -------
=== RUN   TestAccAppConfigurationKeysDataSource_allkeys
=== PAUSE TestAccAppConfigurationKeysDataSource_allkeys
=== CONT  TestAccAppConfigurationKeysDataSource_allkeys
testcase.go:110: Step 1/1 error: Error running apply: exit status 1
Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230118072400473047/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230118072400473047" / Label "label1" / Key "key1": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
with azurerm_app_configuration_key.test,
on terraform_plugin_test.tf line 44, in resource "azurerm_app_configuration_key" "test":
44: resource "azurerm_app_configuration_key" "test" {
checking for presence of existing Key: Configuration Store Id
"/subscriptions/*******/resourceGroups/acctestRG-appconfig-230118072400473047/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230118072400473047"
/ Label "label1" / Key "key1": appconfiguration.BaseClient#GetKeyValue:
Failure responding to request: StatusCode=403 -- Original Error:
autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
EOF
Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230118072400473047/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230118072400473047" / Label "label2" / Key "key1": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
with azurerm_app_configuration_key.test2,
on terraform_plugin_test.tf line 52, in resource "azurerm_app_configuration_key" "test2":
52: resource "azurerm_app_configuration_key" "test2" {
checking for presence of existing Key: Configuration Store Id
"/subscriptions/*******/resourceGroups/acctestRG-appconfig-230118072400473047/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230118072400473047"
/ Label "label2" / Key "key1": appconfiguration.BaseClient#GetKeyValue:
Failure responding to request: StatusCode=403 -- Original Error:
autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
EOF
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Still have test failures:

```
------- Stdout: -------
=== RUN   TestAccAppConfigurationKeysDataSource_allkeys
=== PAUSE TestAccAppConfigurationKeysDataSource_allkeys
=== CONT  TestAccAppConfigurationKeysDataSource_allkeys
    testcase.go:110: Step 1/1 error: Error running apply: exit status 1
        
        Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873" / Label "label1" / Key "key1": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
        
          with azurerm_app_configuration_key.test,
          on terraform_plugin_test.tf line 44, in resource "azurerm_app_configuration_key" "test":
          44: resource "azurerm_app_configuration_key" "test" {
        
        checking for presence of existing Key: Configuration Store Id
        "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873"
        / Label "label1" / Key "key1": appconfiguration.BaseClient#GetKeyValue:
        Failure responding to request: StatusCode=403 -- Original Error:
        autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
        EOF
        
        Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873" / Label "label2" / Key "key1": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
        
          with azurerm_app_configuration_key.test2,
          on terraform_plugin_test.tf line 52, in resource "azurerm_app_configuration_key" "test2":
          52: resource "azurerm_app_configuration_key" "test2" {
        
        checking for presence of existing Key: Configuration Store Id
        "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873"
        / Label "label2" / Key "key1": appconfiguration.BaseClient#GetKeyValue:
        Failure responding to request: StatusCode=403 -- Original Error:
        autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
        EOF
        
        Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873" / Label "testlabel" / Key "key2": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
        
          with azurerm_app_configuration_key.test3,
          on terraform_plugin_test.tf line 60, in resource "azurerm_app_configuration_key" "test3":
          60: resource "azurerm_app_configuration_key" "test3" {
        
        checking for presence of existing Key: Configuration Store Id
        "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873"
        / Label "testlabel" / Key "key2": appconfiguration.BaseClient#GetKeyValue:
        Failure responding to request: StatusCode=403 -- Original Error:
        autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
        EOF
        
        Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873" / Label "testlabel" / Key "key3": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
        
          with azurerm_app_configuration_key.test4,
          on terraform_plugin_test.tf line 68, in resource "azurerm_app_configuration_key" "test4":
          68: resource "azurerm_app_configuration_key" "test4" {
        
        checking for presence of existing Key: Configuration Store Id
        "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873"
        / Label "testlabel" / Key "key3": appconfiguration.BaseClient#GetKeyValue:
        Failure responding to request: StatusCode=403 -- Original Error:
        autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error:
        EOF
        
        Error: checking for presence of existing Key: Configuration Store Id "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873" / Label "" / Key "key3": appconfiguration.BaseClient#GetKeyValue: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
        
          with azurerm_app_configuration_key.test5,
          on terraform_plugin_test.tf line 76, in resource "azurerm_app_configuration_key" "test5":
          76: resource "azurerm_app_configuration_key" "test5" {
        
        checking for presence of existing Key: Configuration Store Id
        "/subscriptions/*******/resourceGroups/acctestRG-appconfig-230130193705070873/providers/Microsoft.AppConfiguration/configurationStores/testacc-appconf230130193705070873"
        / Label "" / Key "key3": appconfiguration.BaseClient#GetKeyValue: Failure
        responding to request: StatusCode=403 -- Original Error: autorest/azure:
        error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
--- FAIL: TestAccAppConfigurationKeysDataSource_allkeys (97.43s)
FAIL
```

**Comment by stephybun** [comment]
@teowa any update on this?

**Comment by teowa** [comment]
The PR should be OK, the failed test is related to https://github.com/hashicorp/terraform-provider-azurerm/pull/20324

**Review by katbyte — CHANGES_REQUESTED** [review]
the tests are still failing:

image

**Comment by brental** [comment]
@stephybun @katbyte @teowa We are very interested in the changes in this PR as we are in a similar situation in regards to assigning permissions. It looks like the unit tests are now passing so would it be possible to progress this?

**Comment by katbyte** [comment]
@brental - while the GH CI is passing the acctests are failing stil.

@teowa - we are still seeing a test failure:

```
------- Stdout: -------
=== RUN   TestAccAppConfiguration_softDeletePurgeThenRecreate
=== PAUSE TestAccAppConfiguration_softDeletePurgeThenRecreate
=== CONT  TestAccAppConfiguration_softDeletePurgeThenRecreate
    testcase.go:110: Step 4/5 error: Error running apply: exit status 1
        
        Error: creating Configuration Store (Subscription: "*******"
        Resource Group Name: "acctestRG-appconfig-230320083720637210"
        Configuration Store Name: "testaccappconf230320083720637210"): polling after Create: Code="NameUnavailable" Message="The specified name is already in use." AdditionalInfo=[{"info":{"activityId":"c857dfe2-fe66-4c04-b2bc-04839b5b19d0"},"type":"ActivityId"}]
        
```

**Comment by teowa** [comment]
HI @stephybun , I have submit another PR https://github.com/hashicorp/terraform-provider-azurerm/pull/21750 to fix the failed test case.

**Comment by ahmar-husain** [comment]
Does anyone know, when this PR will be merged?

File: internal/services/appconfiguration/app_configuration_resource.go [hunk 3]
```diff
@@ -679,3 +666,15 @@ func parsePublicNetworkAccess(input string) *configurationstores.PublicNetworkAc
...
+
+func userIsMissingNecessaryPermission(name, location string) string {
+	return fmt.Sprintf(`
+An existing soft-deleted App Configuration exists with the Name %q in the location %q, however
+the credentials Terraform is using has insufficient permissions to check for an existing soft-deleted App Configuration.
+You can opt out of this behaviour by using the "features" block (located within the "provider" block) - more information
+can be found here:
+https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#features
+Alternatively you can manually recover this (e.g. using the Azure CLI) and then import
+this into Terraform via "terraform import".
```
Comment (tombuildsstuff) [inline internal/services/appconfiguration/app_configuration_resource.go:678]: but the user wouldn't have permission to do that (recovering via the CLI)?

**Review by stephybun — CHANGES_REQUESTED** [review]
@teowa looks like a rebase is required
```
------- Stdout: -------
=== RUN   TestAccAppConfiguration_softDeletePurgeThenRecreate
=== PAUSE TestAccAppConfiguration_softDeletePurgeThenRecreate
=== CONT  TestAccAppConfiguration_softDeletePurgeThenRecreate
testcase.go:110: Step 4/5 error: Error running apply: exit status 1
Error: creating Configuration Store (Subscription: "*******"
Resource Group Name: "acctestRG-appconfig-230601085823840487"
Configuration Store Name: "testaccappconf230601085823840487"): polling after Create: internal-error: a polling status of `Failed` should be surfaced as a PollingFailedError
with azurerm_app_configuration.test,
on terraform_plugin_test.tf line 26, in resource "azurerm_app_configuration" "test":
26: resource "azurerm_app_configuration" "test" {
--- FAIL: TestAccAppConfiguration_softDeletePurgeThenRecreate (659.92s)
FAIL
```

**Comment by teowa** [comment]
> @teowa looks like a rebase is required
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccAppConfiguration_softDeletePurgeThenRecreate
> === PAUSE TestAccAppConfiguration_softDeletePurgeThenRecreate
> === CONT  TestAccAppConfiguration_softDeletePurgeThenRecreate
> testcase.go:110: Step 4/5 error: Error running apply: exit status 1
> Error: creating Configuration Store (Subscription: "*******"
> Resource Group Name: "acctestRG-appconfig-230601085823840487"
> Configuration Store Name: "testaccappconf230601085823840487"): polling after Create: internal-error: a polling status of `Failed` should be surfaced as a PollingFailedError
> with azurerm_app_configuration.test,
> on terraform_plugin_test.tf line 26, in resource "azurerm_app_configuration" "test":
> 26: resource "azurerm_app_configuration" "test" {
> --- FAIL: TestAccAppConfiguration_softDeletePurgeThenRecreate (659.92s)
> FAIL
> ```

This should be resolved in https://github.com/hashicorp/go-azure-sdk/pull/479

**Review by stephybun — APPROVED** [review]
Tests are passing now, thanks @teowa LGTM 👍

**Pull request merged**